Add initial version of the OPA SDK

tsandall opened the pull request:
See the commit messages for details. We should update the docs to include a section in the `Integrating OPA` page once we have done an integration or two to test out the APIs.

anderseknert commented on sdk/opa.go:
```diff
@@ -0,0 +1,405 @@
+			rego.Query(query),
+			rego.Compiler(opa.manager.GetCompiler()),
+			rego.Store(opa.manager.Store),
+			rego.Transaction(txn),
```
What does it mean that we hold a reference to a transaction (and I guess other per-query specific references like metrics) given that this query will be cached for further use? The transaction stored here won't be the one used in any other queries except for the first.. so I'm guessing this is unused?

anderseknert replied:
Just confusing to me that we store stale snapshot data like the transaction and the query metrics on a `rego` object which is reused for each query. If I looked at the object in a debugger I would assume that those values were there because they were needed/used.

> If we don't provide one, the rego package will open one anyway. There's no point in opening two so just provide it here.

I don't get it :) If the transaction stored is stale, won't the rego package have to open one anyway?

anderseknert reviewed (dismissed):
I guess I'll need to step up my game with my competing SDK PR now 😆 

LGTM with a few comments and questions.

anderseknert commented on docs/content/management-discovery.md:
```diff
@@ -52,19 +52,18 @@ to OPA is referred to as the "boot configuration".
-    keyid: my_global_key
-    scope: read
+  resource: /configuration/example/discovery.tar.gz
+signing:
```
Right, this config looks wrong. I'd just remove the signing part.

tsandall replied:
Removed.

srenatus commented on sdk/opa.go:
```diff
@@ -0,0 +1,405 @@
+	}
+
+	if !opa.async {
+		if err := opa.waitForReadiness(ctx); err != nil {
```
[nit] I guess we could inline this. It's a bit of a letdown to scroll all the way down to waitForReadiness to find that one select 😉

tsandall replied:
Fixed.

srenatus commented on sdk/opa.go:
```diff
@@ -0,0 +1,405 @@
+	defer opa.mtx.Unlock()
+
+	result, ok := opa.preparedQueries[query]
+	if !ok {
```
[nit] I'd flip these... but it's a minor style preference only
```go
result, ok := opa.preparedQueries[query]
if ok {
    return result, nil
}

pq, err := rego.New(....)
```

srenatus commented on sdk/opa_test.go:
```diff
@@ -0,0 +1,302 @@
+
+	defer opa.Stop(ctx)
+
+	select {
+	case <-readyCh:
+		// ready
+	}
```
```suggestion
<-readyCh:  // ready
```

unless I'm missing something 🤔

tsandall replied:
Not missing anything. I was on autopilot.

srenatus commented on sdk/opa_test.go:
```diff
@@ -0,0 +1,302 @@
+	defer cancel()
+
+	_, err := New(ctx, Config(config))
+	if err == nil || err.Error() != "context deadline exceeded" {
```
```suggestion
	if !errors.Is(err, context.DeadlineExceeded) {
```
(⚠️ untested, sorry)

tsandall replied:
LGTM. Fixed.

anderseknert commented on plugins/plugins.go:
```diff
@@ -743,11 +749,16 @@ func (m *Manager) Services() []string {
 
-// Logger gets the logger implementation associated with this plugin manager
-func (m *Manager) Logger() sdk.Logger {
+// Logger gets the debug logger for this plugin manager.
```
> Adding "default" or "standard" doesn't really clarify the intent. 

Default or standard clarifies intent in that it's the logger people know and recognize from OPA, not some special logger for SDK debugging (though it can certainly be used for that).

> Also, the logging APIs are leveled by design, so we wouldn't have a "debug level logger"...

"debug level logger" == a logger set to debug level

```go
logger := NewStandardLogger()
logger.SetLevel(Debug)
```

tsandall replied:
Did the s/debug/standard/ replacement!

srenatus commented on plugins/bundle/config.go:
```diff
@@ -172,8 +173,20 @@ func (c *Config) validateAndInjectDefaults(services []string, keys map[string]*k
+
+		if strings.HasPrefix(source.Resource, "file://") {
+			if u, err := url.Parse(source.Resource); err != nil || u.Scheme != "file" {
+				return fmt.Errorf("invalid URL for bundle %q: %v", name, err)
```
Something is odd here, isn't it? We first check that the prefix is `file://`, and then check the `u.Scheme` again? I've been wondering because if `err` is ni and `u.Scheme` is **not** `"file"`, the error would look a bit funny:
```
invalid URL for bundle xyz: nil
```
...but I cannot come up with a `source.Resource` string that would trigger that. 🤔

tsandall replied:
Agreed. The scheme will always be file because of the prefix check. Removed the second condtion.

srenatus commented on plugins/bundle/config_test.go:
```diff
@@ -340,6 +338,14 @@ b3:
 			resource: "/some/longer/path/bundle.tar.gz",
 		},
+		"b4": {
+			service:  "s1",
```
Why would this be `s1`? 🤔 Is that the default because no service was provided in [the config below `b4`](https://github.com/open-policy-agent/opa/pull/3429/commits/7bfb804975d4022b5723d4f3f22a8af1bd32d129#diff-f1d9625ec0db3e9c4594fab73d5880b226c8ead819efcf98bdceaf9a3b89bccfR313)?

tsandall replied:
The service will default to the first one in the list.

srenatus reviewed (approved):
LGTM just nitpicks and questions 👍

tsandall commented on plugins/bundle/plugin.go:
```diff
@@ -585,3 +599,46 @@ func (p *Plugin) getBundlePersistPath() (string, error) {
+	}()
+}
+
+func (fl *fileLoader) Stop(context.Context) {
```
Fixed.

tsandall commented on logging/test/test.go:
```diff
@@ -0,0 +1,92 @@
+	Message string
+}
+
+// Logger logging implementation that does nothing
```
Fixed.

tsandall commented:
OK, I think I've addressed all the feedback and/or replied to the comments. @srenatus if you can do another review of the sdk/opa.go implementation that would be helpful, particularly the new `Configure()` API that supports in-place configuration updates.

srenatus commented on sdk/opa.go:
```diff
@@ -0,0 +1,426 @@
+			}
+		}
+
+		close(ready)
+		done = true
```
I think `ready` and `done` might be redundant?

```go
select {
  case <-ready:
    return
  default: // don't block
}
```
should do the same trick as
```go
if done {
  return
}
```

tsandall replied:
Yep. And now we don't need the extra mutex either.

srenatus commented on sdk/opa.go:
```diff
@@ -0,0 +1,426 @@
+	}
+
+	if block {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ready:
+			break
+		}
```
```suggestion
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ready:
		}
```
Works the same, I think?

tsandall replied:
👍

srenatus commented on sdk/opa.go:
```diff
@@ -0,0 +1,426 @@
+	opa.mtx.Lock()
+	defer opa.mtx.Unlock()
+
+	// NOTE(tsandall): there is no return value from Stop() and it could block
```
Sounds like a TODO, doesn't it? If configured to block, I'd expect all required operations to have finished before it proceeds 💭

tsandall replied:
Added an extra TODO about potential for manual triggers here.

srenatus commented on sdk/opa_test.go:
```diff
@@ -0,0 +1,496 @@
+	}
+}
+
+func TestConfigure(t *testing.T) {
```
leaktest could be useful here, since multiple go routines are involved. Or not. But we could try it.

👉 https://github.com/open-policy-agent/opa/blob/82cca627647165804f8cf042cfd7f84dd13ddecc/rego/rego_wasmtarget_test.go#L202

srenatus commented on sdk/opa.go:
```diff
@@ -0,0 +1,426 @@
+	return opa.configure(ctx, opts.config, opts.Ready, opts.block)
+}
+
+func (opa *OPA) configure(ctx context.Context, bs []byte, ready chan struct{}, block bool) error {
```
[nit] I guess if it's only for closing, `<-chan struct{}` would do, too?

srenatus replied:
"cannot close a receive-only channel" -- not sure what I was thinking last night :D

srenatus commented on sdk/options.go:
```diff
@@ -0,0 +1,100 @@
+	// Ready sets a channel to notify when the OPA instance is ready. If this
+	// field is not set, the Configure() function will block until ready. The
+	// channel is closed to signal readiness.
+	Ready chan struct{}
```
🤔 I guess equivalently, we could allow the caller to provide a function handle which we'd call on ready. (Block if it's nil.) But I rather like this :D

srenatus commented on sdk/options.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+// ConfigOptions contains parameters to (re-)configure OPA.
+type ConfigOptions struct {
```
https://blog.golang.org/module-compatibility#TOC_5. Could be a relevant read... but I don't think we worry too much about comparability of this struct...

srenatus reviewed (approved):
I've re-reviewed sdk/opa.go and sdk/opa_test.go ✔️ 

Just nit-pick level comments! 🙃
